pkg/pillar/cmd/volumemgr/handleverifier.go

```diff
@@ -266,70 +340,30 @@ func handleVerifyImageStatusDelete(ctxArg interface{}, key string,
 	log.Infof("handleVerifyImageStatusDelete done for %s", key)
 }
 
-func lookupPersistImageStatus(ctx *volumemgrContext, objType string,
-	imageSha string) *types.PersistImageStatus {
-
-	if imageSha == "" {
-		return nil
-	}
-	sub := ctx.subscription(types.PersistImageStatus{}, objType)
-	s, _ := sub.Get(imageSha)
-	if s == nil {
-		log.Infof("lookupPersistImageStatus(%s) not found for %s", imageSha, objType)
-		return nil
-	}
-	status := s.(types.PersistImageStatus)
-	return &status
-}
-
-func handlePersistImageStatusModify(ctxArg interface{}, key string,
-	statusArg interface{}) {
-
-	status := statusArg.(types.PersistImageStatus)
-	ctx := ctxArg.(*volumemgrContext)
-
-	log.Infof("handlePersistImageStatusModify for sha: %s, "+
-		" RefCount %d Expired %t", status.ImageSha256, status.RefCount,
-		status.Expired)
+// ImageDownloadDirNames - Returns pendingDirname, verifierDirname, verifiedDirname
+// for the image.
+func ImageDownloadDirNames(objType, ImageID, ImageSha256 string) (string, string, string) {
```

Comment:
These funcs - `ImageDownloadDirNames()` and `ImageDownloadFilenames()` - are nearly identical replicas of what is in `VerifyImageStatus`. Beyond the code-repetition issue, I think we should be able to get rid of the ones on `VerifyImageStatus`. It is only ever consumed by verifier (I just did a search), and one of the key goals of this PR is to make the verifier "dumber": it gets a path to a file, is asked to verify it, and returns true or false. Thus, it should never have access to anything other than "path to the file to verify".

Since you (correctly) moved the process of moving a file from `pending/` to `verifier/` to `verified/` into volumemgr, there isn't need for anything else to know it.

I think we can remove that entire chunk from the `VerifyImageStatus` (which makes it unlikely something else will abuse it accidentally), and just have two simple fields on the `VerifyImageStatus`:

* `Target` (`Path` or anything you want): this is where verifier will look to find the file it should verify
* `Final` (or whatever naming you want): this is where volumemgr has placed the final verified file

We only have the second one because we use the `VerifyImageStatus` to indicate that a file has been fully verified. If we don't need it - for example, if, once verified, and after volumemgr has moved the verified file to its final location, we do not use `VerifyImageStatus` again - then we do not need the second one at all. I just am not sure of that.

Either way, we should get rid of all of that from `VerifyImageStatus` so that it is not duplicate, and so that nothing ever is tempted to abuse it, and so that there is a very clear, explicit, volumemgr setting the target on the `VerifyImageConfig` that the verifier consumes.

Reply:
> Since you (correctly) moved the process of moving a file from `pending/` to `verifier/` to `verified/` into volumemgr, there isn't need for anything else to know it.

@deitch As I commented on the document, we don't want to do the second move in volumemgr if we want to enable verifier being the only service (future container) with write access to the verified directories. But volumemgr can tell the verifier which Target pathname to put the file in (and verifier can check that it is in one of its verified directories). Thus we can have multiple files per directory etc.

Reply:
@eriknordmark that makes it somewhat dependent on location. But I think what you are saying is, you want to have a directory (or volume, or whatever) that verifier is the only thing that can write to it. Anything else is blocked. It never will be foolproof - container volumes are just bind-mounted, anything with root or in dom0 eventually can get in - but it does provide an additional layer.

Other things might tell verifier one of:

* verify the file at location X
* delete a file

But other things cannot modify files in verifier's "personal space".

If that is correct, then I would _not_ have the volumemgr tell verifier where to save the files. Instead, I would have verifier own its own "verified" directory (configurable on startup). That directory would _never_ be exposed via `types.` or some public property.

The flow would look like this:

1. volumemgr needs an image, tells downloader, "download it to path X"
1. downloader downloads to X
1. volumemgr needs to verify the image, tells verifier, "verify the file at path X"
1. verifier tries to verify the file:
    * failure: report failure, do nothing with file
    * success: move it somewhere under verifiers' own control, report success
1. volumemgr consumes that file, knowing that the specific file in the specific directory is under control of verifier alone

If volumemgr instead tells it: "verify file at location X and when done move it to Y", then verifier's output is under control of anything that can send verifier a message.

Does that meet both needs?

Reply:
@deitch You have to move the file before running the verification, otherwise the caller can race and modify the file after verification and before the move. 

I think it would be useful for the volumemgr to be able to specify at least the basename of the file; it might not need to specify the full pathname. That enables it to have multiple files in flight without separate directories. But that is a detail.

Reply:
> You have to move the file before running the verification

I think you are saying that the following could happen:

1. volumemgr tells downloader to download to path X
1. downloader downloads to path X
1. volumemgr moves file from path X to path Y
1. volumemgr tells verifier to verify file at path Y
1. verifier verifies at path Y
1. verifier moves file to path Z (protected path)
1. verifier tells volumemgr that verified file is at Z

Between the 3rd-to-last to 2nd-to-last stages, after verification but before moving file from Y to Z (protected), something could modify the file.

I think it is a low risk, but I agree that it could happen. If we are being so strict as to separate these services, then we should address it.

But that is something for _verifier_ to handle. It is within its responsibility to:

1. Receive a file
1. Verify it
1. Confirm it

If, to do that, it needs to do an extra move, let it:

1. volumemgr tells downloader to download to path X
1. downloader downloads to path X
1. volumemgr moves file from path X to path Y
1. volumemgr tells verifier to verify file at path Y
1. verifier moves file from Y to an internally (to verifier) controlled and protected cache path Y2
1. verifier verifies at path Y2
1. verifier moves file to path Z (protected path)
1. verifier tells volumemgr that verified file is at Z

The simpler form of that is:

1. volumemgr tells downloader to download to path X
1. downloader downloads to path X
1. volumemgr tells verifier to verify file at path X
1. verifier moves file from X to an internally (to verifier) controlled and protected cache path Y
1. verifier verifies at path Y
1. verifier moves file to path Z (protected path)
1. verifier tells volumemgr that verified file is at Z

In current lingo:

* volumemgr owns and manages `download/`
* volumemgr does no file moving; tells verifier to verify file in `download/`
* verifier owns both `verifier/` and `verified/`, but these are internal to them completely
* verifier moves file from `download/` to `verifier/` before verifying, and from `verifier/` to `verified/` after verifying
* verifier tells volumemgr, "your verified file is at `verified/`

From volumemgr's perspective, there is no such thing as `downloads/` or `verifier/` or `verified/`. There is just

* where I tell downloader to place downloads
* where I tell verifier the input file to a verification exists
* where verifier tells me a verified file exists

Reply:
Updated verifier to read the downloaded file location from config. VolumeMgr will be unaware of the verifier and verified dir. After Verifier verifies the file, the final verified location is published to volumeMgr.